### PR TITLE
docs(router): Updated usage of observables in tutorial and dev guide

### LINKED ISSUE #2675: Routing docs should use idiomatic Observable syntax
Ref: https://angular.io/docs/ts/latest/tutorial/toh-pt5.html

current example:

``` ts
ngOnInit(): void {
  this.route.params.forEach((params: Params) => {
    let id = +params['id'];
    this.heroService.getHero(id)
      .then(hero => this.hero = hero);
  });
}
```

This is rather hard to parse (and not exactly side effect free)

Better syntax would be 

``` ts
this.routeParams.flatMap(params => this.heroService.getHero(params.id))
```

or

``` ts
this.routeParams.switchMap(params => this.heroService.getHero(params.id))
```

which would cancel in flight requests - the current syntax has the possibility of an inflight request arriving after the user navigates away.

Note that .switch/flatMap can accept a Promise or an Observable, so its a good place to show interop.

### PR DESCRIPTION
**Pull request by brandonroberts**
Closes #2675 

Updated usage of observable route parameters and updated prose
Moved route configurations into separate variable for consistency
Added async pipe to handle subscriptions for list items
Removed old routing files

cc: @wardbell @robwormald

### CODE REVIEW

File: public/docs/ts/latest/guide/router.jade [hunk 1]
```diff
@@ -1096,8 +1112,14 @@ code-example(language="bash").
 +makeExcerpt('app/heroes/hero-list.component.ts (import)', 'import-router')
 
 :marked
+  We'll import the `switchMap` operator to perform operations on our `Observable`
+  of route parameters.
+
++makeExcerpt('app/heroes/hero-list.component.ts (rxjs imports)', 'rxjs-imports')
+
+:marked
   Then we use the `ActivatedRoute` to access the `params` _Observable_ so we can subscribe
```
Comment (wardbell) [inline public/docs/ts/latest/guide/router.jade]: This sentence (which I probably wrote originally) needs to be reworked. Too wooden, too complicated.

Try:
--
Then we inject the `ActivatedRoute` in the `HeroListComponent` constructor.

[show the code snippet]

The  `ActivatedRoute.params` property is an `Observable` of route parameters. The `params` emits new `id` values when the user navigates to the component.  In `ngOnInit` we subscribe to those values, set the `selectedId`, and get the heroes. 
--
Also re-title the code snippet which purports to be about the constructor but is actually about both the constructor and the `ngOnInit`.

File: public/docs/ts/latest/tutorial/toh-pt5.jade [hunk 2]
```diff
@@ -530,14 +530,27 @@ block ngOnInit
...
+    while our hero is being retrieved.
 
 - var _str2int = _docsFor == 'dart' ? '<code>int.parse</code> static method' : 'JavaScript (+) operator'
 :marked
   The hero `id` is a number. Route parameters are *always strings*.
   So we convert the route parameter value to a number with the !{_str2int}.
 
+.l-sub-section
+  :marked
+    ### Do I need to unsubscribe?
```
Comment (wardbell) [inline public/docs/ts/latest/tutorial/toh-pt5.jade]: Thanks for adding this.

File: public/docs/ts/latest/tutorial/toh-pt5.jade [hunk 3]
```diff
@@ -530,14 +530,27 @@ block ngOnInit
 
 block extract-id
   :marked
-    Notice how we extract the `id` by calling the `forEach` method
-    which will deliver our !{_array} of route parameters.
+    Notice how we use the `switchMap` operator on the `Observable` to
```
Comment (wardbell) [inline public/docs/ts/latest/tutorial/toh-pt5.jade]: I wonder why the original said there is an array of route parameters. There is a params _object_ and the  `+params['id']` expression gets the `id` parameter from that object. Weird.

Rewrite
...
Note how the `switchMap` operator maps the `id` in the observable route parameters
to a new `Observable`, the result of the  `HeroService.getHero` method.

If the user re-navigates to this component while a `getHero` request is still inflight,
`switchMap` cancels that old request before calling  `HeroService.getHero` again.
...

File: public/docs/ts/latest/guide/router.jade [hunk 4]
```diff
@@ -884,13 +889,24 @@ a#hero-detail-ctor
...
 +makeExcerpt('app/heroes/hero-detail.component.ts (ngOnInit)', 'ngOnInit')
 
 .l-sub-section
   :marked
+    The `switchMap` operator allows you to perform an action with the current value of the `Observable`,
+    and map it to a new `Observable`. As with many `rxjs` operators, `switchMap` handles
+    an `Observable` as well as a `Promise` to retrieve the value they emit.
+
+    We could have used the `flatMap` operator, which works the same way with one exception.
+    The `switchMap` operator will also cancel any in-flight requests if our user re-navigates to the route
```
Comment (wardbell) [inline public/docs/ts/latest/guide/router.jade]: Upon further reflection, while I remain profoundly uncomfortable with what seems like a distraction to me, I'm letting it slide.

File: public/docs/_examples/toh-5/ts/app/dashboard.component.ts [hunk 5]
```diff
@@ -33,7 +33,7 @@ export class DashboardComponent implements OnInit {
 
   ngOnInit(): void {
     this.heroService.getHeroes()
-      .then(heroes => this.heroes = heroes.slice(1, 5));
+      .then((heroes: Hero[]) => this.heroes = heroes.slice(1, 5));
```
Comment (brandonroberts) [inline public/docs/_examples/toh-5/ts/app/dashboard.component.ts]: My editor was giving me type errors. I thought something changed with the linting. I'll leave it as-is

File: public/docs/_examples/toh-5/ts/app/dashboard.component.2.ts [hunk 6]
```diff
@@ -20,7 +20,7 @@ export class DashboardComponent implements OnInit {
 
   ngOnInit(): void {
     this.heroService.getHeroes()
-      .then(heroes => this.heroes = heroes.slice(1, 5));
+      .then((heroes: Hero[]) => this.heroes = heroes.slice(1, 5));
```
Comment (brandonroberts) [inline public/docs/_examples/toh-5/ts/app/dashboard.component.2.ts]: Reverted

File: public/docs/_examples/toh-5/ts/app/heroes.component.ts [hunk 7]
```diff
@@ -29,7 +29,7 @@ export class HeroesComponent implements OnInit {
     private heroService: HeroService) { }
 
   getHeroes(): void {
-    this.heroService.getHeroes().then(heroes => this.heroes = heroes);
+    this.heroService.getHeroes().then((heroes: Hero[]) => this.heroes = heroes);
```
Comment (brandonroberts) [inline public/docs/_examples/toh-5/ts/app/heroes.component.ts]: Reverted

File: public/docs/ts/latest/guide/router.jade [hunk 8]
```diff
@@ -884,13 +889,24 @@ a#hero-detail-ctor
...
 
 +makeExcerpt('app/heroes/hero-detail.component.ts (ngOnInit)', 'ngOnInit')
 
 .l-sub-section
   :marked
+    The `switchMap` operator allows you to perform an action with the current value of the `Observable`,
+    and map it to a new `Observable`. As with many `rxjs` operators, `switchMap` handles
+    an `Observable` as well as a `Promise` to retrieve the value they emit.
+
+    We could have used the `flatMap` operator, which works the same way with one exception.
```
Comment (brandonroberts) [inline public/docs/ts/latest/guide/router.jade]: Removed  `flatMap` sentence.

**Comment by brandonroberts** [comment]
@wardbell I made the suggest changes to the prose for the usage of switchMap and other concerns.

**Comment by wardbell** [comment]
Excellent. Will merge right after #2718 with Igor's suggestion. Want to make sure no merge conflicts.

**Comment by wardbell** [comment]
Resolved merge conflicts (all in `toh-pt5.jade`) since the last review. Merging as soon as it clears Travis
